keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,386 @@
+(if node3 is capable).
+
+### Risks and Mitigations
+
+The feature requires additional processing for pods that use it and it is ok to
+have some performance overhead. But we will make sure our implementation will
+not have any performance penalty for pods that do not use this feature.
+
+## Design Details
+
+### API
+
+A new structure called `EvenSpreading` is introduced which acts as a standalone
+spec and applies to `pod.spec`. It's only effective when it's not nil.
+
+```go
+type PodSpec struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
+
+#### Option 1
+
+Inside `EvenSpreading`, we need hard affinityTerms (similar with
+`PodAffinityTerm`) and soft affinityTerms (similar with
+`WeightedPodAffinityTerm`). This describes when we perform even distribution,
+which pods are considered as a group.
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
+}
+```
+
+#### Option 2
+
+Another option is to flatten "required" and "perferred" podAffinityTerms, and
+eliminate embedded "TopologyKey":
+
+```go
+type PodSelectorTerm struct {
+	  // A label query over a set of resources, in this case pods.
+	  // +optional
+	  LabelSelector *metav1.LabelSelector
+	  // namespaces specifies which namespaces the labelSelector applies to (matches against);
+	  // null or empty list means "this pod's namespace"
+	  // +optional
+	  Namespaces []string
+}
+
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // ImbalanceToleratable indicates whether pods imbalance across TopologyKey is toleratable
+    // - false (default) denotes it's not toleratable and internaly it's a Predicate (hard requirement)
+    // - true denotes it's toleratable and internally it's a Priority (soft requirement)
+    // Note that only pods skew beyonds "MaxSkew" is considered as imbalanced
+    ImbalanceToleratable bool
+    // PodSelector defines a series of podSelectorTerm and applied at scheduling time.
+    PodSelector []PodSelectorTerm
```

Comment:
`PodSelectors`? `SiblingPods`? `PodsInGroup`? `SpreadingGroup`?

What happens if different pods have different notions of the group?

Reply:
Updated to use a single `labelSelector`.